File: #98 Validate Binary Search Tree/98_validate_binary_search_tree.py
"""

---
MEDIUM
98. Validate Binary Search Tree
---

Given the root of a binary tree, determine if it is a valid binary search tree (BST).

A valid BST is defined as follows:

The left 
subtree
 of a node contains only nodes with keys less than the node's key.
The right subtree of a node contains only nodes with keys greater than the node's key.
Both the left and right subtrees must also be binary search trees.
 

Example 1:


Input: root = [2,1,3]
Output: true
Example 2:


Input: root = [5,1,4,null,null,3,6]
Output: false
Explanation: The root node's value is 5 but its right child's value is 4.
 

Constraints:

The number of nodes in the tree is in the range [1, 104].
-231 <= Node.val <= 231 - 1

"""

## Solution 1
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def valid(self, node, minimum, maximum):
        if not node:
            return True

        if not (minimum < node.val < maximum):
            return False

        return self.valid(node.left, minimum, node.val) and self.valid(node.right, node.val, maximum)

    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.valid(root, float('-inf'), float('inf'))

## Solution 2
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorder(self, root, nodes):
        if not root:
            return

        self.inorder(root.left, nodes)
        nodes.append(root.val)
        self.inorder(root.right, nodes)

    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True

        nodes = []
        # Get the in-order traversal of the tree
        self.inorder(root, nodes)

        # Check if there are duplicates
        if len(nodes) != len(set(nodes)):
            return False

        # Check if the list is sorted in ascending order
        for i in range(len(nodes) - 1):
            if nodes[i] >= nodes[i + 1]:
                return False

        return True

File: #98 Validate Binary Search Tree/98_validate_binary_search_tree.cpp
/*

---
MEDIUM
98. Validate Binary Search Tree
---

Given the root of a binary tree, determine if it is a valid binary search tree (BST).

A valid BST is defined as follows:

The left 
subtree
 of a node contains only nodes with keys less than the node's key.
The right subtree of a node contains only nodes with keys greater than the node's key.
Both the left and right subtrees must also be binary search trees.
 

Example 1:


Input: root = [2,1,3]
Output: true
Example 2:


Input: root = [5,1,4,null,null,3,6]
Output: false
Explanation: The root node's value is 5 but its right child's value is 4.
 

Constraints:

The number of nodes in the tree is in the range [1, 104].
-231 <= Node.val <= 231 - 1

*/

// Solution 1
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool valid(TreeNode* node, long minimum, long maximum) {
        if (!node) return true;

        if (!(node->val > minimum && node->val < maximum)) return false;

        return valid(node->left, minimum, node->val) && valid(node->right, node->val, maximum);
    }

    bool isValidBST(TreeNode* root) {
        return valid(root, LONG_MIN, LONG_MAX);
    }
};

// Solution 2
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void inorder(TreeNode* root, vector<int> &nodes) {
        if(!root) return;

        inorder(root->left, nodes);
        nodes.push_back(root->val);    
        inorder(root->right, nodes); 
    }

    bool isValidBST(TreeNode* root) {
        if(!root) return true;

        vector <int> nodes;
        inorder(root, nodes);   // Get the in-order traversal of the tree 

        set<int> s(nodes.begin(), nodes.end());

        if(s.size() != nodes.size()) return false;  // Check if there are duplicates

        // Check if sorted in ascending order
        for(int i = 0; i < nodes.size()-1; i++) {
            if(nodes[i] > nodes[i+1]) return false;
        }

        return true;
    }
};